### DISCUSSION TITLE
Is 2FA possible?

### DISCUSSION THREAD

**Discussion by frederikhors** [maxcountryman/axum-login#134] [Q&A]
Is 2FA possible?

**Comment by maxcountryman** [comment]
It would be helpful if you could provide a little more context around your question.

The short answer is `axum-login` is intentionally designed in a way that aims to not limit you as much as is practical. And so yes, multi-factor authentication can be built into an application that uses this crate.

**Reply by frederikhors** [reply]
Thanks for your answer.

A little background:

Now in my app I have an authentication session created using `Keycloack` and at each API call I first verify that the user is authenticated and authorized (on Keycloack server) and then continue with my code.

As you can imagine maintaining Keycloack is both cumbersome and slow for my APIs and I would like a more streamlined solution like `axum-login`.

The only thing I don't understand is how to integrate 2FA systems into `axum-login`.

Can you show me the way?

Should I use crates like https://crates.io/crates/totp and https://crates.io/crates/totp-rs?

What do you think about adding an example on https://github.com/maxcountryman/axum-login/tree/main/examples/sqlite?

**Reply by maxcountryman** [reply]
At a very high level, MFA can be incorporated into the authentication flow however you might like.

Depending on what you're doing exactly, this might be entirely independent of axum-login itself (altho if you were using a third party to verify MFA status on each request, you could incorporate that into the `get_user` method implementation).

What I'm getting at here is a common way of doing this in a web application is to check MFA in one step and then with a successful check perform a separate authentication step (such as username and password). So axum-login doesn't have a specific opinion about that. If it's appropriate to combine that into a single step then you can implement `authenticate` such that it takes the MFA state as part of the `Credentials` type and do the verification within that implementation.

If you need help with a specific app I could offer to help directly--I would charge a rate for an engagement like this.
